Sources/_MatchingEngine/Regex/Parse/Mocking.swift

```diff
@@ -1,107 +1,142 @@
 
-/// Mock-up of the Swift lexer, so we can model our interfaces
-struct SwiftLexer {
-  enum Tok {
-    case opaque(AST)
-
-    var ast: AST {
-      switch self {
-      case .opaque(let ast): return ast
-      }
-    }
-  }
+// TODO: mock up multi-line soon
 
-  static func lex(_ src: String) -> Tok {
-    // Assume there's just a single literal here, no escaping
-    if let start = src.firstIndex(of: "'") {
-      return lexSingleQuotedLiteral(
-        src[src.index(after: start)..<src.lastIndex(of: "'")!])
+public enum Delimiter: Hashable, CaseIterable {
+  case regular
+  case modern
+
+  var openingAndClosing: (opening: String, closing: String) {
+    switch self {
+    case .regular: return ("'/", "/'")
+    case .modern: return ("'|", "|'")
     }
-    fatalError()
   }
-
-  static func lexSingleQuotedLiteral(_ sub: Substring) -> Tok {
-    let ast: AST
-    if sub.first == "/" && sub.last == "/" {
-      ast = try! parse(
-        sub.dropFirst().dropLast(), .traditional)
-    } else {
-      ast = try! parse(sub, .modern)
+  public var opening: String { openingAndClosing.opening }
+  public var closing: String { openingAndClosing.closing }
+
+  /// The default set of syntax options that the delimiter indicates.
+  public var defaultSyntaxOptions: SyntaxOptions {
+    switch self {
+    case .regular: return .traditional
+    case .modern: return .modern
     }
+  }
+}
 
-    // TODO: ... call regex parser
-    return .opaque(ast)
+public struct LexError: Error, CustomStringConvertible {
+  enum Kind: Hashable {
+    case endOfString
+    case invalidUTF8 // TODO: better range reporting
+    case unknownDelimiter
   }
 
-}
+  var kind: Kind
 
-// TODO: mock up multi-line soon
+  /// The pointer at which to resume lexing.
+  public var resumePtr: UnsafeRawPointer
 
+  init(_ kind: Kind, resumeAt resumePtr: UnsafeRawPointer) {
+    self.kind = kind
+    self.resumePtr = resumePtr
+  }
 
-enum LexError: Error {
-  case endOfString
-  case invalidUTF8 // TODO: better range reporting
-  case notRegexLiteralStart
+  /// Whether the lexer came across an unknown delimiter, indicating that this
+  /// is not in fact a regex literal.
+  public var isUnknownDelimiter: Bool { kind == .unknownDelimiter }
+
+  public var description: String {
+    switch kind {
+    case .endOfString: return "unterminated regex literal"
+    case .invalidUTF8: return "invalid UTF-8 found in source file"
+    case .unknownDelimiter: return "unknown regex literal delimiter"
+    }
+  }
 }
 
+/// Drop a set of regex delimiters from the input string, returning the contents
+/// and the delimiter used. The input string must have valid delimiters.
+func droppingRegexDelimiters(_ str: String) -> (String, Delimiter) {
```

Comment:
I don't understand the point of this function, why would we have a parallel implementation over String?

Reply:
It's used by both:

- The runtime to drop the delimiters off and get the syntax options to use
- The C++ parser as we don't currently track the contents range on the C++ `Token` so we need to drop off the delimiters again when we come to parse

We could have both of those go back through the lexing function, I don't feel too strongly about it. I agree it's not ideal having two implementations of effectively the same operation